### PR TITLE
ambient handle pod cleanups

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
can you add tests or explain what the PR is for?

File: pilot/pkg/serviceregistry/kube/controller/ambientindex.go [hunk 1]
```diff
@@ -124,23 +125,20 @@ func (a *AmbientIndex) updateWaypoint(sa model.WaypointScope, ipStr string, isDe
 			}
 			wl.WaypointAddresses = addrs
 			if filtered {
-				// If there was a change, also update the VIPs and record for a push
+				// If there was a change, record for a push
 				updates[model.ConfigKey{Kind: kind.Address, Name: wl.ResourceName()}] = struct{}{}
-				a.byPod[wl.ResourceName()] = wl
-				for vip := range wl.VirtualIps {
-					a.insertWorkloadToService(vip, wl)
-				}
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambientindex.go:132]: For this i think we donot need to insert again, it is done in handlePod cb

File: pilot/pkg/serviceregistry/kube/controller/ambientindex.go [hunk 2]
```diff
@@ -149,14 +147,9 @@ func (a *AmbientIndex) updateWaypoint(sa model.WaypointScope, ipStr string, isDe
 				}
 			}
 			if !found {
-				wl := wl.Clone()
 				wl.WaypointAddresses = append(wl.WaypointAddresses, addr)
 				// If there was a change, also update the VIPs and record for a push
 				updates[model.ConfigKey{Kind: kind.Address, Name: wl.ResourceName()}] = struct{}{}
-				a.byPod[wl.ResourceName()] = wl
-				for vip := range wl.VirtualIps {
-					a.insertWorkloadToService(vip, wl)
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambientindex.go:158]: same as above

**Comment by hzxuzhonghu** [comment]
@howardjohn There is `TestAmbientIndex` there.

File: pilot/pkg/serviceregistry/kube/controller/ambientindex.go [hunk 3]
```diff
@@ -591,24 +587,28 @@ func (c *Controller) setupIndex() *AmbientIndex {
...
+			scope := model.WaypointScope{Namespace: p.Namespace, ServiceAccount: p.Annotations[constants.WaypointServiceAccount]}
 			ip := p.Status.PodIP
 			if isDelete || !IsPodReady(p) {
-				if idx.waypoints[n].Contains(ip) {
-					idx.waypoints[n].Delete(ip)
-					updates.Merge(idx.updateWaypoint(n, ip, true, c))
+				if idx.waypoints[scope].Contains(ip) {
+					idx.waypoints[scope].Delete(ip)
+					if len(idx.waypoints[scope]) == 0 {
+						delete(idx.waypoints, scope)
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambientindex.go:602]: make sure no leak here

**Comment by howardjohn** [comment]
> @howardjohn There is `TestAmbientIndex` there.

So its just refactor? no behavior change? The PR description sounds like it is fixing a bug but the tests are not changed

File: pilot/pkg/serviceregistry/kube/controller/ambientindex.go (hunk 3 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go:602]: nit: use sets.DeleteCleanupLast

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go:602]: and sets.InsertOrNew below

File: pilot/pkg/serviceregistry/kube/controller/ambientindex.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go:132]: How does this work? We get a Pod update for the waypoint which calls handlePod, but we needto update the user pods

**Review by howardjohn — COMMENTED** [review]
LGTM aside from my one comment which I think is why the test fails

File: pilot/pkg/serviceregistry/kube/controller/ambientindex.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go:125]: In service handler, wl has been updated,

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go:125]: But this is for waypoint pod updating, there is no Service involved.

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go:125]: `a.insertWorkloadToService(vip, wl)` is actually insert user pod to `byService` right?

When a waypoint updates, it has no influence on it

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go:125]: ping @howardjohn

**Review by howardjohn — APPROVED** [review]

**Pull request merged**